gazebo_plugins/src/gazebo_ros_joint_effort.cpp

```diff
@@ -0,0 +1,144 @@
+  gazebo::physics::ModelPtr model_;
+
+  /// Joints being tracked.
+  std::vector<gazebo::physics::JointPtr> joints_;
+
+  /// Joint state subscriber
+  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_state_sub_;
+
+  /// Container for the effort that this plugin exerts on the body.
+  sensor_msgs::msg::JointState joint_state_msg_;
+
+  // Pointer to the update event connection
+  gazebo::event::ConnectionPtr update_connection_;
+};
+
+GazeboRosJointEffort::GazeboRosJointEffort()
+: impl_(std::make_unique<GazeboRosJointEffortPrivate>())
+{
+}
+
+GazeboRosJointEffort::~GazeboRosJointEffort()
+{
+}
+
+void GazeboRosJointEffort::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  impl_->model_ = model;
+
+  // ROS node
+  impl_->ros_node_ = gazebo_ros::Node::Get(sdf);
+
+  // Joints
+  if (!sdf->HasElement("joint_name")) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(), "Plugin missing <joint_name>s");
+    impl_->ros_node_.reset();
+    return;
+  }
+
+  sdf::ElementPtr joint_elem = sdf->GetElement("joint_name");
+  while (joint_elem) {
+    auto joint_name = joint_elem->Get<std::string>();
+
+    auto joint = model->GetJoint(joint_name);
+    if (!joint) {
+      RCLCPP_ERROR(impl_->ros_node_->get_logger(), "Joint %s does not exist!", joint_name.c_str());
+    } else {
+      impl_->joints_.push_back(joint);
+    }
+    joint_elem = joint_elem->GetNextElement("joint_name");
+  }
+
+  if (impl_->joints_.empty()) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(), "No joints found.");
+    impl_->ros_node_.reset();
+    return;
+  }
+
+  impl_->joint_state_sub_ = impl_->ros_node_->create_subscription<sensor_msgs::msg::JointState>(
+    "joint_efforts", rclcpp::SystemDefaultsQoS(),
+    std::bind(&GazeboRosJointEffort::OnRosJointStateMsg, this, std::placeholders::_1));
+
+  // Callback on every iteration
+  impl_->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin(
+    std::bind(&GazeboRosJointEffort::OnUpdate, this));
+
+  impl_->joint_state_msg_.effort.resize(impl_->joints_.size());
+}
+
+void GazeboRosJointEffort::OnRosJointStateMsg(const sensor_msgs::msg::JointState::SharedPtr msg)
+{
+  if(msg->name.empty()) {
+      RCLCPP_ERROR(
+        impl_->ros_node_->get_logger(),
+        "No joint names in msg");
+        return;
+  }
+  for (size_t i = 0; i < msg->effort.size(); ++i){
+    if (msg->name[i].empty()) {
+      RCLCPP_ERROR(
+        impl_->ros_node_->get_logger(),
+        "Joint %d with empty name",
+        i);
+      impl_->joint_state_msg_.effort.clear();
+      return;
+    }
+    auto joint = impl_->model_->GetJoint(msg->name[i]);
+    if (!joint) {
+      RCLCPP_ERROR(
+        impl_->ros_node_->get_logger(),
+        "Joint %s not found in SDF!",
+        msg->name[i].c_str());
+        return;
+    }
+    impl_->joint_state_msg_.effort[i] = msg->effort[i];
+  }
+  impl_->joint_state_msg_.header = msg->header;
+}
+
+void GazeboRosJointEffort::OnUpdate()
+{
```

Comment:
If you copy the whole message as I mentioned before, you could do the name checking here instead, so there's no need to keep track of indices at all. You'd do something like:

```.cpp
void GazeboRosJointEffort::OnUpdate()
{
  if (impl_->joint_state_msg_.name.size() != impl_->joint_state_msg_.effort.size()) {
    // print error
    return;
  }

  for (size_t i = 0; i < impl_->joint_state_msg_.name.size(); ++i){
    auto joint = impl_->model_->GetJoint(impl_->joint_state_msg_.name[i]);
    if (joint)
      joint->SetForce(0, impl_->joint_state_msg_.effort[i]);
  }
}
```

Reply:
I've tried to implement this way but `effort.size()` is different than `name.size()`. Because of that, I've also added a case to check whether `name` is empty.  Thank you for your suggestions and your heads up concerning the IMU commit

Reply:
> effort.size() is different than name.size()

It sounds like the message may be malformed? The [spec](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/JointState.html) says that _"All arrays in this message should have the same size, or be empty"_. I believe the name is the only one which shouldn't be empty, because _"Each joint is uniquely identified by its name"_.

Reply:
It makes sense. In your opinion, which conditions of malformed msgs should be verified?